File: cli/src/commands/validate.ts
/**
 * mrsf validate — Validate one or all sidecar files.
 */

import type { Command } from "commander";
import chalk from "chalk";
import { validateFile } from "../lib/validator.js";
import { resolveSidecarPaths } from "../lib/resolve-files.js";
import type { ValidationResult } from "../lib/types.js";

export function registerValidate(program: Command): void {
  program
    .command("validate [files...]")
    .description("Validate MRSF sidecar files against the schema and spec rules")
    .option("-s, --strict", "Treat warnings as errors")
    .action(async (files: string[], opts: { strict?: boolean }) => {
      const parentOpts = program.opts();
      const cwd = parentOpts.cwd ?? process.cwd();
      const quiet = parentOpts.quiet ?? false;
      const verbose = parentOpts.verbose ?? false;

      const sidecarPaths = await resolveSidecarPaths(files, cwd);

      if (sidecarPaths.length === 0) {
        if (!quiet) console.log("No sidecar files found.");
        return;
      }

      let hasErrors = false;

      for (const sp of sidecarPaths) {
        let result: ValidationResult;
        try {
          result = await validateFile(sp);
        } catch (err: unknown) {
          const msg = err instanceof Error ? err.message : String(err);
          console.error(chalk.red(`✗ ${sp}: ${msg}`));
          hasErrors = true;
          continue;
        }

        const effectiveValid = opts.strict
          ? result.valid && result.warnings.length === 0
          : result.valid;

        if (!effectiveValid) hasErrors = true;

        if (!quiet || !effectiveValid) {
          const icon = effectiveValid ? chalk.green("✓") : chalk.red("✗");
          console.log(`${icon} ${sp}`);
        }

        for (const e of result.errors) {
          console.log(chalk.red(`  ERROR: ${e.message}${e.path ? ` (${e.path})` : ""}`));
        }
        for (const w of result.warnings) {
          const fn = opts.strict ? chalk.red : chalk.yellow;
          console.log(fn(`  ${opts.strict ? "ERROR" : "WARN"}: ${w.message}${w.path ? ` (${w.path})` : ""}`));
        }
        if (verbose) {
          console.log(chalk.dim(`  ${result.errors.length} error(s), ${result.warnings.length} warning(s)`));
        }
      }

      if (hasErrors) process.exit(1);
    });
}

File: python/src/mrsf/cli/validate.py
"""mrsf validate [files...] — validate sidecar files."""

from __future__ import annotations

import sys

import click

from ..resolve_files import resolve_sidecar_paths
from ..validator import validate_file
from .main import MrsfContext, pass_ctx


@click.command("validate")
@click.argument("files", nargs=-1, type=click.Path())
@click.option("-s", "--strict", is_flag=True, help="Treat warnings as errors.")
@pass_ctx
def validate_cmd(ctx: MrsfContext, files: tuple[str, ...], strict: bool) -> None:
    """Validate sidecar files against the MRSF schema."""
    sidecars = resolve_sidecar_paths(list(files), cwd=ctx.cwd)
    if not sidecars:
        click.echo("No sidecar files found.", err=True)
        sys.exit(1)

    has_errors = False
    for sc in sidecars:
        result = validate_file(sc)
        ok = result.valid
        errors = result.errors
        warnings = result.warnings

        if strict and warnings:
            ok = False

        if ok:
            if not ctx.quiet:
                click.echo(click.style(f"✓ {sc}", fg="green"))
        else:
            has_errors = True
            click.echo(click.style(f"✗ {sc}", fg="red"))

        for err in errors:
            path_suffix = f" ({err.path})" if err.path else ""
            click.echo(click.style(f"  ERROR: {err.message}{path_suffix}", fg="red"))
        for warn in warnings:
            style = "red" if strict else "yellow"
            path_suffix = f" ({warn.path})" if warn.path else ""
            click.echo(click.style(f"  WARN: {warn.message}{path_suffix}", fg=style))

    if has_errors:
        sys.exit(1)
